Funcky/Extensions/EnumerableExtensions/AdjacentGroupBy.cs

```diff
@@ -0,0 +1,219 @@
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements in <paramref name="source"/> sequence.</typeparam>
+        /// <typeparam name="TKey">The type of the key returned by keySelector.</typeparam>
+        /// <typeparam name="TElement">The type of the elements in each IGrouping{TKey,TElement}.</typeparam>
+        /// <typeparam name="TResult">The type of the result value returned by resultSelector.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="keySelector">A function to extract the key for each element.</param>
+        /// <param name="elementSelector">A function to map each source element to an element in the IGrouping{TKey,TElement}.</param>
+        /// <param name="resultSelector">A function to create a result value from each group.</param>
+        /// <param name="comparer">An IEqualityComparer{T} to compare keys.</param>
+        /// <returns>A collection of elements of type TResult where each element represents a projection over a group and its key.</returns>
+        [Pure]
+        public static IEnumerable<TResult> AdjacentGroupBy<TSource, TKey, TElement, TResult>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TSource, TElement> elementSelector,
+            Func<TKey, IReadOnlyList<TElement>, TResult> resultSelector,
+            IEqualityComparer<TKey> comparer)
+        {
+            using var enumerator = source.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                yield break;
+            }
+
+            var (group, key) = CreateGroupAndKey(keySelector, elementSelector, enumerator);
+
+            while (enumerator.MoveNext())
+            {
+                if (comparer.Equals(key, keySelector(enumerator.Current)))
+                {
+                    group = group.Add(elementSelector(enumerator.Current));
+                }
+                else
+                {
+                    yield return resultSelector(key, group);
+                    (group, key) = CreateGroupAndKey(keySelector, elementSelector, enumerator);
+                }
+            }
+
+            yield return resultSelector(key, group);
+        }
+
+        internal static Grouping<TKey, TElement> CreateGrouping<TKey, TElement>(TKey key, IReadOnlyList<TElement> elements)
+            => new Grouping<TKey, TElement>(key, elements);
+
+        private static (ImmutableList<TElement> Group, TKey Key) CreateGroupAndKey<TSource, TKey, TElement>(Func<TSource, TKey> keySelector, Func<TSource, TElement> elementSelector, IEnumerator<TSource> enumerator)
+        {
+            var group = ImmutableList.Create(elementSelector(enumerator.Current));
+            var key = keySelector(enumerator.Current);
+
+            return (group, key);
+        }
+
+        internal class Grouping<TKey, TElement> : IGrouping<TKey, TElement>, IReadOnlyList<TElement>
```

Comment:
I'd re-add the `IList<T>` "implementation", since some of the LINQ operators have an optimization for `IList<T>`:
* [`First`](https://github.com/dotnet/runtime/blob/e7204f5d6fcaca5e097ec854b3be6055229fc442/src/libraries/System.Linq/src/System/Linq/First.cs)
* [`Last`](https://github.com/dotnet/runtime/blob/e7204f5d6fcaca5e097ec854b3be6055229fc442/src/libraries/System.Linq/src/System/Linq/Last.cs)